onset_maproom/calc.py

```diff
@@ -33,6 +35,123 @@ def onset_date(daily_rain, early_start_day, early_start_month, search_days, rain
 #  onset_date = onset_date[time_coord] + onset_date  
   return onset_date
 
+# Time functions
+
+def daily_tobegroupedby_season(daily_data, start_day, start_month, end_day, end_month, time_coord="T"):
+  """Returns dataset ready to be grouped by with:
+  the daily data where all days not in season of interest are dropped
+  season_starts: an array where the non-dropped days are indexed by the first day of their season -- to use to groupby
+  seasons_ends: an array with the dates of the end of the seasons
+  Can then apply groupby on daily_data against seasons_starts, and preserving seasons_ends for the record
+  If starting day-month is 29-Feb, uses 1-Mar.
+  If ending day-month is 29-Feb, uses 1-Mar and triggers the option to open the right edge of the Interval.
+  That means that the last day included in the season will be 29-Feb in leap years and 28-Feb otherwise
+  """
+  #Deal with leap year cases
+  if start_day == 29 and start_month == 2 :
+    start_day = 1
+    start_month = 3
+  end_day2 = end_day
+  end_month2 = end_month
+  if end_day == 29 and end_month == 2 :
+    end_day2 = 1
+    end_month2 = 3
+  #Drop date outside very first and very last edges -- this ensures we get complete seasons with regards to edges, later on
+  start_edges = daily_data[time_coord].where(
+    ((daily_data[time_coord].dt.day==start_day) & (daily_data[time_coord].dt.month==start_month)),
+    drop=True
+  )
+  end_edges = daily_data[time_coord].where(
+    ((daily_data[time_coord].dt.day==end_day2) & (daily_data[time_coord].dt.month==end_month2)),
+    drop=True
+  )
+  daily_data = daily_data.sel(**{time_coord: slice(start_edges[0],end_edges[-1])})
+  start_edges = daily_data[time_coord].where(
+    ((daily_data[time_coord].dt.day==start_day) & (daily_data[time_coord].dt.month==start_month)),
+    drop=True
+  )
+  end_edges = daily_data[time_coord].where(
+    ((daily_data[time_coord].dt.day==end_day2) & (daily_data[time_coord].dt.month==end_month2)),
+    drop=True
+  )
+  #Creates array of edges of the season that will form the bins
+  seasons_edges = xr.concat([start_edges, end_edges], "T_out", join="override")
+  #Creates seasons_starts that will be used for grouping
+  #and seasons_ends that is one of the outputs
+  if end_day == 29 and end_month == 2 :
+    days_in_season = (
+      daily_data[time_coord] >= seasons_edges.isel(T_out=0).rename({time_coord: "group"})
+    ) & (
+      daily_data[time_coord] < seasons_edges.isel(T_out=1).rename({time_coord: "group"})
+    )
+  else:
+    days_in_season = (
+      daily_data[time_coord] >= seasons_edges.isel(T_out=0).rename({time_coord: "group"})
+    ) & (
+      daily_data[time_coord] <= seasons_edges.isel(T_out=1).rename({time_coord: "group"})
+    )
+  seasons_starts = daily_data[time_coord].where(days_in_season)
```

Comment:
it makes a 2D array of dates values that depend on the full time T dimension and the n groups (4 years here). There are date values where the date is within the right group/year and within the seasons edges. And groups has the advantage of being indexed by the day that starts the season. Thus, I can use where to replace, in each groups, the non NaT values by the value of group dimension (the day that starts that season); and then use min (or max really) to squeeze the groups dimension that has no more purpose now and get back to a 1D that depends on T and where all days that are within a season have for value the first day of the season. And I can use that to groupby so that my groupedby dimension is the one I want: the first day of each season, instead of meaningless number 1 to n.
So basically the telescoping from 1D to 2D allows me to treat the groups of days within each season differently, as I need to.

Reply:
I sort of half understand. Is there documentation that says what it means to use a 2D array to select elements from a 1D array, or is this just obvious to people who are used to working with array broadcasting?

I imagine there's another way to do it that (a) is easier for me to understand and (b) doesn't require a quadratic amount of memory, but we don't need to find it right now. If this works, let's go with it.

Reply:
where mentions broadcasting quite a bit. I am 100% sure what it means though so not sure I interpret correctly.
http://xarray.pydata.org/en/stable/generated/xarray.DataArray.where.html#xarray.DataArray.where
https://numpy.org/doc/stable/reference/generated/numpy.where.html#numpy.where

I did that intuitively though...

Another way is to return to the cumsum trick then the group values [1...12...23...34...4...] turn out to be the positions (-1) of the dates I want in their place from start_edges. Is there a way to replace them by transitivity or composition? something like seasons_start = start_edges([1...12...23...34...4...])? I think I tried that but failed but that may just be my python clumsiness. So if you tell me there should be a way, I can try again.

Reply:
I am onto something

Reply:
another way found and implemented in last commit

Reply:
OK, I understand this version!